File: scenarios/scala/scala-binary-parser/template/tests/TestBinaryParser.scala
import java.nio.ByteBuffer
import java.nio.ByteOrder

object TestBinaryParser {
  def main(args: Array[String]): Unit = {
    try {
      testLogin()
      testOrder()
      println("All tests passed.")
    } catch {
      case e: Throwable => 
        println(s"FAILED: ${e.getMessage}")
        e.printStackTrace()
        System.exit(1)
    }
  }

  def testLogin(): Unit = {
    val magic = 0xAF.toByte
    val version = 1.toByte
    val startType = 1.toByte
    val userLen = 5.toByte
    val userBytes = "admin".getBytes("UTF-8")
    val hash = Array.fill[Byte](32)(0xAA.toByte)
    
    val payload = Array(magic, version, startType, userLen) ++ userBytes ++ hash
    val sum = payload.map(_.toInt & 0xFF).sum
    val checksum = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(sum).array()
    
    val data = payload ++ checksum
    val result = BinaryParser.parsePacket(data)
    
    assert(result("type") == "Login")
    assert(result("username") == "admin")
    assert(result("hash").asInstanceOf[Array[Byte]].length == 32)
    println("Test Login: PASSED")
  }

  def testOrder(): Unit = {
    val magic = 0xAF.toByte
    val version = 1.toByte
    val tType = 2.toByte
    
    val id = 123456789012345L
    val symbol = "AAPL".getBytes("UTF-8")
    val side = 1.toByte
    val price = 150.50
    val qty = 100
    
    val bb = ByteBuffer.allocate(8 + 4 + 1 + 8 + 4).order(ByteOrder.BIG_ENDIAN)
    bb.putLong(id)
    bb.put(symbol)
    bb.put(side)
    bb.putDouble(price)
    bb.putInt(qty)
    val body = bb.array()
    
    val payload = Array(magic, version, tType) ++ body
    val sum = payload.map(_.toInt & 0xFF).sum
    val checksum = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(sum).array()
    
    val result = BinaryParser.parsePacket(payload ++ checksum)
    
    assert(result("type") == "Order")
    assert(result("id") == id)
    assert(new String(result("symbol").asInstanceOf[Array[Byte]]) == "AAPL" || result("symbol") == "AAPL")
    assert(Math.abs(result("price").asInstanceOf[Double] - 150.50) < 0.001)
    println("Test Order: PASSED")
  }
}

File: scenarios/ruby/rb-binary-parser/template/tests/test_binary_parser.rb
require_relative '../src/binary_parser'
require 'test/unit'

class TestBinaryParser < Test::Unit::TestCase
  def test_login
    magic = "\xAF"
    version = "\x01"
    type = "\x01"
    user_len = "\x05"
    username = "admin"
    hash = "\xAA" * 32
    
    payload = magic + version + type + user_len + username + hash
    
    sum = payload.each_byte.reduce(0, :+)
    checksum = [sum].pack("N")
    
    data = payload + checksum
    
    parser = BinaryParser.new
    result = parser.parse_packet(data)
    
    assert_equal 'Login', result[:type]
    assert_equal 'admin', result[:username]
    assert_equal 32, result[:hash].bytesize
  end

  def test_order
    magic = "\xAF"
    version = "\x01"
    type = "\x02"
    
    id = [123456789012345].pack("Q>") # Q> is 64-bit BE
    symbol = "AAPL"
    side = "\x01"
    price = [150.50].pack("G") # G is Double BE
    qty = [100].pack("N")
    
    payload = magic + version + type + id + symbol + side + price + qty
    
    sum = payload.each_byte.reduce(0, :+)
    checksum = [sum].pack("N")
    
    parser = BinaryParser.new
    result = parser.parse_packet(payload + checksum)
    
    assert_equal 'Order', result[:type]
    assert_equal 123456789012345, result[:id]
    assert_equal 'AAPL', result[:symbol]
    assert_in_delta 150.50, result[:price], 0.001
  end
end
